docs/libvirt-pod-networking.md

```diff
@@ -0,0 +1,82 @@
+# Using CNI Interfaces and libvirt for networking
+
+Author: Fabian Deutsch \<fabiand@redhat.com\>
+
+## Introduction
+
+Today networking in KubeVirt is not integrated with Kubernetes, a user can
+rather choose to create a specific VM Spec definition to create a particular
+libvirt domxml, to reuse an existing physical device.
+But this is obviously suboptimal
+
+- It is not integrated with Kubernetes
+- It does not provide connectivity with pods
+- It requires to specify host specific bits (the network device to use)
+
+The purpose of this document is to describe one approach of how we can provide
+a minimal integrated solution, with few workarounds.
+
+
+## Use-case
+
+The primary use-case is to allow VM to get connected to the pod network.
+
+
+## API
+
+The API is pretty clear: As usual an interface will be defined.
+To indicate that Pod interfaces shall be used for networking, a new type
+`pod-network` is introduced.
+All other fields, concerning the properties of the virtual NIC, can be used
+as usual.
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: pod-network
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+        source:
+          network: default
+```
+
+The `source` field can be omitted initially, as Kubernetes only provides a
+single network for Pods today.
+In future we might want to allow it to select the Pod network to connect to.
+
+
+## Implementation
+
+The broader steps to enable networking in the described way:
+
+1. Request/Create a new Pod interface on the libvirt pod
+2. Remember the IP of the Pod interface, then remove it from the Pod interface
+3. Create a libvirt network, backed by a bridge, enable DHCP,
+   and add a single host entry, with the IP address of the Pod, and the MAC
+   address of the corresponding virtual NIC of the VM
```

Comment:
Ok, so IIUC, as an example consider the libvirt POD has eth0 with IP  10.0.0.2, and the VM POD has its own eth0 with IP  10.0.0.3.  We've put the eth0 from the libvirt POD into a bridge, and attaching a QEMU guest TAP device to that bridge.  Now we're running DHCP on this bridge so that it hands out the IP address associated with the VM POD's eth0 to the new QEMU tap device. Finally we've unconfigured the eth0 in the VM POD, so it no longer has an address.

In theory this can work, but I think there is significant risk that this will break. In particular, I would expect k8s network layer to have some kind of IP address & MAC address anti-spoofing protection.  ie, k8s network would prevent the libvirt POD's eth0, from sending/receiving any packets with address 10.0.0.3, because that address belongs to a different POD.

The second complication arises if we want to expose some network services from within the VM POD. eg if we wanted to expose SPICE client from the VM POD. You can't have the NIC in the VM POD used both for host services like SPICE, and for guest OS usage. Host management network needs to be separate from the guest OS network. 

So the question is whether the IP address associated with a VM POD is best thought of as a guest NIC IP address, or as the address to use for exposing QEMU network services like SPICE / serial ports / migration traffic / etc.

My other PR to move QEMU into the libvirtd network namespace, instead of a separate network namespace, suggests an answer to the question. ie that the VM POD IP address is exclusively to use as the IP for a guest virtual NIC, and that QEMU network services all live on the IP address associated with libvirtd.